src/ranged.cpp

```diff
@@ -617,11 +563,12 @@ int player::fire_gun( const tripoint &target, int shots, item& gun )
         }
     }
 
-    practice( skill_gun, train_skill ? 15 : 0 );
-
     // Use different amounts of time depending on the type of gun and our skill
     moves -= time_to_fire( *this, *gun.type );
 
+    practice( skill_gun, xp * ( get_skill_level( skill_gun ) + 1 ) );
```

Comment:
Why the linearization of skill gain?
The `xp` is implicitly linear with skill anyway, it could be a better idea to allow skill gain to scale with square of range (divided by constant). For example, that idea with soft gain for low ranges above, except extended to the whole range. If that's too much, there could be an upper cap, like `MAX_SKILL * get_skill_level( marks )`.